N°3806 If php memory_limit is lower than 256M, set 256M. If it's greater do …

BGdu38 opened the pull request:
My memory_limit is set to 512M, but for csv import memory limit is hard coded in iTop.

BGdu38 committed:
If php memory_limit is lower than 256M, set 256M. If it's greater do nothing.

Hipska reviewed (changes requested):
iTop already has a helper method for that, so the code should be like:
```php
if (utils::ConvertToBytes(ini_get('memory_limit')) < '268435456') {
```
But it should be moved to after the includes in that case.

Question to combodo: Should `ini_set` actually happen?

piRGoif commented:
Hello,
Many thanks for this contribution !
This is a known limitation, that is referenced in Combodo internal db as N°3806. The fix is planned for 2.7.5 but as low priority, so it might not make its way to this version ? We'll see in the upcoming weeks as the development has just started.
I'm putting this PR on hold for now, as this may not be the solution that will be chosen ?

Molkobain commented:
> iTop already has a helper method for that, so the code should be like:
> 
> ```
> if (utils::ConvertToBytes(ini_get('memory_limit')) < '268435456') {
> ```

Indeed using the `utils::ConvertToBytes()` method is indeed a good practice, good catch.

> But it should be moved to after the includes in that case.

Not sure to follow? It was already before them.

Hipska commented:
> > But it should be moved to after the includes in that case.
> 
> Not sure to follow? It was already before them.

Well, the helper method needs to be loaded in php before it can be used. Currently the `ini_set` is before all the loading.

Molkobain commented:
> > Not sure to follow? It was already before them.
> 
> Well, the helper method needs to be loaded in php before it can be used. Currently the `ini_set` is before all the loading.

Right! I thought you were saying the `ini_set` itself had to be before.

piRGoif commented:
Using utils::ConvertToBytes would be a good idea indeed ! But Hipska you're right, we need to mind to call it AFTER the required calls, whereas the ini_set is currently before them.

Searching for other memory_limit occurrences, I found all of those ini_set calls : 
* csvimport.php of course O:)
* pages/ajax.render.php (operation "xlsx_run")
* setup/ajax.dataloader.php
* webservices/backoffice.dataloader.php

I will reference this in our internal ticket !

Also, in the setup (\SetupUtils::CheckPhpAndExtensions) we have a call to `\utils::IsMemoryLimitOk`, I think this should be the one to call !

piRGoif commented:
Also this PR is based and develop... We will certainly do a cherry pick manually to integrate it :/

piRGoif committed:
Add generic method to set memory_limit

New method : \utils::SetMinMemoryLimit

Remplace ini_set with new method for csvimport and XLSX export.
Didn't change 2 ini_set calls in setup (setup/ajax.dataloader.php and webservices/backoffice.dataloader.php) as the new method doesn't offer exactly the same services, and aren't generating the same bug.

piRGoif commented:
I've just wrote a proposition, I've contacted @BGdu38, awaiting for his answer.

Hipska commented on application/utils.inc.php:
```diff
@@ -658,48 +658,86 @@ public static function GetUploadTempId($sTransactionId = null)
+		if (!file_exists($sFileName)) {
+			return false;
+		}
+
```
Seems not relevant for this PR?

piRGoif replied:
You're right ! The formatter usually does this kind of things, modifying only modified lines... but just a few extra more. That was the case here also, it just added some formatting on existing code just above modified lines.

Hipska commented on application/utils.inc.php:
```diff
@@ -658,48 +658,86 @@ public static function GetUploadTempId($sTransactionId = null)
+	/**
+	 * Checks if the memory limit is at least what is required
+	 *
+	 * @param int $memoryLimit set limit in bytes
+	 * @param int $requiredLimit required limit in bytes
+	 *
+	 * @return bool
+	 */
+	public static function IsMemoryLimitOk($memoryLimit, $requiredLimit)
+	{
+		return ($memoryLimit >= $requiredLimit) || ($memoryLimit == -1);
+	}
```
```suggestion
	 * @param int $iMemoryLimit set limit in bytes
	 * @param int $iRequiredLimit required limit in bytes
	 *
	 * @return bool
	 */
	public static function IsMemoryLimitOk($iMemoryLimit, $iRequiredLimit)
	{
		return ($iMemoryLimit >= $iRequiredLimit) || ($iMemoryLimit == -1);
	}
```

Hipska commented on application/utils.inc.php:
```diff
@@ -658,48 +658,86 @@ public static function GetUploadTempId($sTransactionId = null)
+		}
+		$iMemoryLimit = utils::ConvertToBytes($sMemoryLimit);
+
+		if (($iMemoryLimit >= $iRequiredLimit) || ($iMemoryLimit === -1)) {
```
```suggestion
		if (static::IsMemoryLimitOk($iMemoryLimit, $iRequiredLimit)) {
```

piRGoif replied:
Wooops !! I feel quite shamy huhu
Thanks, good catch !

piRGoif replied:
Also havin a strict comparison with -1 is a good think, I'll add it in another commit

Hipska commented on test/application/UtilsTest.php:
```diff
@@ -552,4 +552,23 @@ public function GetMentionedObjectsFromTextProvider(): array
+	public function ConvertToBytesProvider()
+	{
+		return [
+			'512M' => ['512M', 512 * 1024 * 1024],
```
Also add test for other options:
```suggestion
			'123' => ['123', 123],
			'56k' => ['56k', 56 * 1024],
			'512M' => ['512M', 512 * 1024 * 1024],
			'2G' => ['2G', 2 * 1024 * 1024 * 1024],
```

piRGoif commented:
I've pushed my modifications.
Comments & reviews are welcome !

piRGoif committed:
:white_check_mark: More items tested in \UtilsTest::ConvertToBytesProvider

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@me.com>

piRGoif committed:
:art: Improve \utils::IsMemoryLimitOk var naming

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@me.com>

piRGoif committed:
Remove code duplicate in \utils::SetMinMemoryLimit

Use \utils::IsMemoryLimitOk instead of duplicating its code !

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@me.com>

piRGoif committed:
:art: Strict comparison + use static:: instead of utils::

piRGoif commented:
Rebased the branch using `git rebase --onto support/2.7 develop PR/215_memory_limit`.
Then changed the base branch for the PR.

Ready to merge !

piRGoif committed:
csvimport : remove use from develop components
Mistake done when rebasing branch from develop to support/2.7, wooops
